File: src/services/routingService.ts
import { RiskForecast, RiskLevel, RiskAssessment } from '../types/risk';
import {
  RoadNode,
  RoadEdge,
  RoadNetworkGraph,
  RoutingMode,
  RouteSegment,
  RouteResult,
  RoutePairResult,
  RouteRiskExposure,
  AvoidedHazardSegment,
} from '../types/routing';
import { PROTOTYPE_ROAD_GRAPH } from '../mock/roadNetwork';

/**
 * Returns penalty multiplier and blocking status for a road edge given its flood conditions and routing mode.
 */
export function calculateEdgeCostMultiplier(
  riskLevel: RiskLevel,
  depth_cm: number,
  mode: RoutingMode,
  isElevated: boolean = false
): { multiplier: number; isBlocked: boolean; reason?: string } {
  // If road is elevated (e.g. Bandra-Worli Sea Link or Eastern Freeway Viaduct),
  // it is raised above surface street water inundation.
  if (isElevated) {
    return { multiplier: mode === 'EMERGENCY' ? 0.85 : 1.0, isBlocked: false };
  }

  // Any street with CRITICAL risk or depth >= 100cm is completely impassable/blocked in all modes
  if (riskLevel === 'CRITICAL' || depth_cm >= 100) {
    return {
      multiplier: Infinity,
      isBlocked: true,
      reason: `Critical flood inundation (${depth_cm.toFixed(1)} cm) - road completely impassable`,
    };
  }

  if (mode === 'EMERGENCY') {
    // High-clearance emergency vehicles (ambulances, fire tenders, rescue trucks)
    // Physical wading limit is 50 cm. Any depth >= 50 cm is blocked for vehicle safety.
    if (depth_cm >= 50) {
      return {
        multiplier: Infinity,
        isBlocked: true,
        reason: `Hazardous water depth (${depth_cm.toFixed(1)} cm) exceeds emergency vehicle wading limit (50 cm)`,
      };
    }
    if (riskLevel === 'VERY_HIGH' || depth_cm >= 30) {
      return { multiplier: 3.5, isBlocked: false };
    }
    if (riskLevel === 'HIGH' || depth_cm >= 20) {
      return { multiplier: 2.0, isBlocked: false };
    }
    if (riskLevel === 'MODERATE' || depth_cm >= 5) {
      return { multiplier: 1.2, isBlocked: false };
    }
    // LOW risk: Priority transit with emergency sirens
    return { multiplier: 0.85, isBlocked: false };
  }

  if (mode === 'SAFEST') {
    // Commuter safety priority: aggressive diversion away from flood waters
    if (depth_cm >= 50) {
      return {
        multiplier: 200.0,
        isBlocked: false,
        reason: `Severe flood depth (${depth_cm.toFixed(1)} cm) - highly penalized`,
      };
    }
    if (riskLevel === 'VERY_HIGH') {
      return { multiplier: 100.0, isBlocked: false };
    }
    if (riskLevel === 'HIGH' || depth_cm >= 20) {
      return { multiplier: 30.0, isBlocked: false };
    }
    if (riskLevel === 'MODERATE' || depth_cm >= 5) {
      return { multiplier: 5.0, isBlocked: false };
    }
    return { multiplier: 1.0, isBlocked: false };
  }

  // FASTEST mode: Commuters balancing travel time and manageable shallow water
  if (depth_cm >= 50) {
    return { multiplier: 15.0, isBlocked: false };
  }
  if (riskLevel === 'VERY_HIGH') {
    return { multiplier: 8.0, isBlocked: false };
  }
  if (riskLevel === 'HIGH' || depth_cm >= 20) {
    return { multiplier: 3.5, isBlocked: false };
  }
  if (riskLevel === 'MODERATE' || depth_cm >= 5) {
    return { multiplier: 1.4, isBlocked: false };
  }
  return { multiplier: 1.0, isBlocked: false };
}

/**
 * Evaluates an edge's current flood status from the horizon's RiskAssessments.
 */
export function getEdgeFloodStatus(
  edge: RoadEdge,
  assessments: RiskAssessment[]
): { depth_cm: number; risk_level: RiskLevel; reason?: string } {
  if (edge.is_elevated) {
    return { depth_cm: 0, risk_level: 'LOW', reason: 'Elevated viaduct / bridge above surface water' };
  }

  const match = assessments.find((a) => a.cellId === edge.catchment_cell_id);
  if (!match) {
    return { depth_cm: 0, risk_level: 'LOW' };
  }

  return {
    depth_cm: Math.max(0, match.depth_cm),
    risk_level: match.riskLevel,
    reason: match.reasons.length > 0 ? match.reasons[0] : undefined,
  };
}

interface DirectedEdge {
  edge: RoadEdge;
  fromNodeId: string;
  toNodeId: string;
  cost: number;
  baseTravelTimeS: number;
  effectiveTravelTimeS: number;
  depth_cm: number;
  risk_level: RiskLevel;
  multiplier: number;
  isBlocked: boolean;
}

/**
 * Builds adjacency map with evaluated flood costs for Dijkstra pathfinding.
 */
function buildAdjacency(
  graph: RoadNetworkGraph,
  assessments: RiskAssessment[],
  mode: RoutingMode,
  penalizedEdgeIds: Set<string> = new Set()
): Map<string, DirectedEdge[]> {
  const adj = new Map<string, DirectedEdge[]>();

  for (const node of graph.nodes) {
    adj.set(node.id, []);
  }

  for (const edge of graph.edges) {
    const flood = getEdgeFloodStatus(edge, assessments);
    const { multiplier, isBlocked } = calculateEdgeCostMultiplier(
      flood.risk_level,
      flood.depth_cm,
      mode,
      edge.is_elevated
    );

    const baseTravelTimeS = edge.distance_m / ((edge.base_speed_kmh * 1000) / 3600);
    const penaltyBonus = penalizedEdgeIds.has(edge.id) ? 2.5 : 1.0;
    const cost = isBlocked ? Infinity : baseTravelTimeS * multiplier * penaltyBonus;
    const effectiveTravelTimeS = isBlocked ? Infinity : baseTravelTimeS * multiplier;

    // Directed transition: source -> target
    const dirA: DirectedEdge = {
      edge,
      fromNodeId: edge.source,
      toNodeId: edge.target,
      cost,
      baseTravelTimeS,
      effectiveTravelTimeS,
      depth_cm: flood.depth_cm,
      risk_level: flood.risk_level,
      multiplier,
      isBlocked,
    };
    adj.get(edge.source)?.push(dirA);

    // If bidirectional, add reverse target -> source
    if (edge.bidirectional) {
      const dirB: DirectedEdge = {
        edge,
        fromNodeId: edge.target,
        toNodeId: edge.source,
        cost,
        baseTravelTimeS,
        effectiveTravelTimeS,
        depth_cm: flood.depth_cm,
        risk_level: flood.risk_level,
        multiplier,
        isBlocked,
      };
      adj.get(edge.target)?.push(dirB);
    }
  }

  return adj;
}

const RISK_RANK: Record<RiskLevel, number> = {
  LOW: 1,
  MODERATE: 2,
  HIGH: 3,
  VERY_HIGH: 4,
  CRITICAL: 5,
};

/**
 * Generates an engineering and commuter explanation for the selected route.
 */
function generateRouteExplanation(
  mode: RoutingMode,
  status: 'FOUND' | 'NO_SAFE_ROUTE' | 'ORIGIN_EQUALS_DESTINATION',
  maxDepthCm: number,
  highestRisk: RiskLevel,
  avoidedSegments: AvoidedHazardSegment[],
  _segments: RouteSegment[]
): string {
  if (status === 'ORIGIN_EQUALS_DESTINATION') {
    return 'Origin and destination are the same location. Zero travel required.';
  }

  if (status === 'NO_SAFE_ROUTE') {
    const reasons = avoidedSegments.slice(0, 3).map((s) => `${s.name} (${s.depth_cm.toFixed(0)}cm, ${s.risk_level})`);
    return `No passable route found under ${mode} navigation criteria. Major corridors are impassable due to severe waterlogging: ${reasons.join(', ')}.`;
  }

  const avoidsText =
    avoidedSegments.length > 0
      ? ` Successfully diverted around ${avoidedSegments.length} flood-prone chokepoint(s) including ${avoidedSegments[0].name} (${avoidedSegments[0].depth_cm.toFixed(0)} cm, ${avoidedSegments[0].risk_level}).`
      : ' All traversed road links maintain dry or low-hazard conditions.';

  if (mode === 'SAFEST') {
    if (maxDepthCm === 0) {
      return `Optimal flood-free route selected.${avoidsText}`;
    }
    return `Safest route prioritized minimal surface water exposure (maximum depth ${maxDepthCm.toFixed(1)} cm, ${highestRisk} hazard).${avoidsText}`;
  }

  if (mode === 'EMERGENCY') {
    return `Emergency response transit corridor optimized for rapid arrival.${avoidsText} Maximum water wading depth along corridor: ${maxDepthCm.toFixed(1)} cm.`;
  }

  return `Fastest transit corridor selected.${avoidsText} Maximum flood depth along route: ${maxDepthCm.toFixed(1)} cm.`;
}

/**
 * Core Dijkstra pathfinding algorithm.
 */
function dijkstra(
  originId: string,
  destinationId: string,
  graph: RoadNetworkGraph,
  assessments: RiskAssessment[],
  mode: RoutingMode,
  penalizedEdgeIds: Set<string> = new Set()
): {
  success: boolean;
  path: DirectedEdge[];
  nodeSequence: string[];
  allAvoidedHazards: AvoidedHazardSegment[];
} {
  const adj = buildAdjacency(graph, assessments, mode, penalizedEdgeIds);

  // Identify network-wide flooded hazard segments for the avoidance report
  const allAvoidedHazards: AvoidedHazardSegment[] = [];
  for (const edge of graph.edges) {
    const flood = getEdgeFloodStatus(edge, assessments);
    if (flood.risk_level === 'HIGH' || flood.risk_level === 'VERY_HIGH' || flood.risk_level === 'CRITICAL') {
      allAvoidedHazards.push({
        edge_id: edge.id,
        name: edge.name,
        risk_level: flood.risk_level,
        depth_cm: flood.depth_cm,
        reason: flood.reason || `Surface flood depth ${flood.depth_cm.toFixed(1)} cm`,
      });
    }
  }

  const dist = new Map<string, number>();
  const prevNode = new Map<string, string>();
  const prevEdge = new Map<string, DirectedEdge>();
  const visited = new Set<string>();

  for (const node of graph.nodes) {
    dist.set(node.id, Infinity);
  }
  dist.set(originId, 0);

  while (visited.size < graph.nodes.length) {
    // Find unvisited node with minimum distance
    let u: string | null = null;
    let minDist = Infinity;

    for (const [nodeId, d] of dist.entries()) {
      if (!visited.has(nodeId) && d < minDist) {
        minDist = d;
        u = nodeId;
      }
    }

    if (!u || minDist === Infinity) break;
    if (u === destinationId) break;

    visited.add(u);

    const neighbors = adj.get(u) || [];
    for (const edge of neighbors) {
      if (visited.has(edge.toNodeId) || edge.isBlocked) continue;

      const alt = dist.get(u)! + edge.cost;
      if (alt < dist.get(edge.toNodeId)!) {
        dist.set(edge.toNodeId, alt);
        prevNode.set(edge.toNodeId, u);
        prevEdge.set(edge.toNodeId, edge);
      }
    }
  }

  if (dist.get(destinationId) === Infinity || !prevNode.has(destinationId)) {
    return {
      success: false,
      path: [],
      nodeSequence: [],
      allAvoidedHazards,
    };
  }

  // Reconstruct path
  const path: DirectedEdge[] = [];
  const nodeSequence: string[] = [destinationId];
  let curr = destinationId;

  while (curr !== originId) {
    const pEdge = prevEdge.get(curr)!;
    path.unshift(pEdge);
    curr = prevNode.get(curr)!;
    nodeSequence.unshift(curr);
  }

  return {
    success: true,
    path,
    nodeSequence,
    allAvoidedHazards,
  };
}

/**
 * Builds a comprehensive RouteResult object from a solved path.
 */
function formatRouteResult(
  origin: RoadNode,
  destination: RoadNode,
  mode: RoutingMode,
  horizon: 'T+0' | 'T+1' | 'T+2' | 'T+3',
  path: DirectedEdge[],
  nodeSequence: string[],
  allAvoidedHazards: AvoidedHazardSegment[],
  nodeMap: Map<string, RoadNode>,
  isAlternative: boolean = false
): RouteResult {
  if (origin.id === destination.id) {
    return {
      status: 'ORIGIN_EQUALS_DESTINATION',
      mode,
      horizon,
      origin,
      destination,
      total_distance_m: 0,
      total_distance_km: 0,
      total_time_s: 0,
      total_time_min: 0,
      max_flood_depth_cm: 0,
      highest_risk_level: 'LOW',
      segments: [],
      path_node_ids: [origin.id],
      coordinates: [[origin.lat, origin.lng]],
      risk_exposure: { low_m: 0, moderate_m: 0, high_m: 0, very_high_m: 0, critical_m: 0 },
      avoided_segments: [],
      explanation: generateRouteExplanation(mode, 'ORIGIN_EQUALS_DESTINATION', 0, 'LOW', [], []),
      is_alternative: isAlternative,
      provenance: 'MODEL OUTPUT / DERIVED',
    };
  }

  if (path.length === 0) {
    return {
      status: 'NO_SAFE_ROUTE',
      mode,
      horizon,
      origin,
      destination,
      total_distance_m: 0,
      total_distance_km: 0,
      total_time_s: 0,
      total_time_min: 0,
      max_flood_depth_cm: 0,
      highest_risk_level: 'CRITICAL',
      segments: [],
      path_node_ids: [],
      coordinates: [],
      risk_exposure: { low_m: 0, moderate_m: 0, high_m: 0, very_high_m: 0, critical_m: 0 },
      avoided_segments: allAvoidedHazards,
      explanation: generateRouteExplanation(mode, 'NO_SAFE_ROUTE', 0, 'CRITICAL', allAvoidedHazards, []),
      is_alternative: isAlternative,
      provenance: 'MODEL OUTPUT / DERIVED',
    };
  }

  let total_distance_m = 0;
  let total_time_s = 0;
  let max_flood_depth_cm = 0;
  let highest_risk_level: RiskLevel = 'LOW';

  const risk_exposure: RouteRiskExposure = {
    low_m: 0,
    moderate_m: 0,
    high_m: 0,
    very_high_m: 0,
    critical_m: 0,
  };

  const segments: RouteSegment[] = [];
  const consolidatedCoordinates: [number, number][] = [];
  const traversedEdgeIds = new Set<string>();

  for (let i = 0; i < path.length; i++) {
    const dEdge = path[i];
    const edge = dEdge.edge;
    traversedEdgeIds.add(edge.id);

    total_distance_m += edge.distance_m;
    total_time_s += dEdge.effectiveTravelTimeS;

    if (dEdge.depth_cm > max_flood_depth_cm) {
      max_flood_depth_cm = dEdge.depth_cm;
    }

    if (RISK_RANK[dEdge.risk_level] > RISK_RANK[highest_risk_level]) {
      highest_risk_level = dEdge.risk_level;
    }

    // Accumulate risk exposure distance
    switch (dEdge.risk_level) {
      case 'LOW':
        risk_exposure.low_m += edge.distance_m;
        break;
      case 'MODERATE':
        risk_exposure.moderate_m += edge.distance_m;
        break;
      case 'HIGH':
        risk_exposure.high_m += edge.distance_m;
        break;
      case 'VERY_HIGH':
        risk_exposure.very_high_m += edge.distance_m;
        break;
      case 'CRITICAL':
        risk_exposure.critical_m += edge.distance_m;
        break;
    }

    // Determine segment coordinates in the correct travel direction
    let segCoords: [number, number][] = [...edge.coordinates];
    if (dEdge.fromNodeId !== edge.source) {
      // Reverse coordinates if traversing against nominal edge direction
      segCoords = [...edge.coordinates].reverse();
    }

    // Append coordinates avoiding duplicate junction points
    for (let j = 0; j < segCoords.length; j++) {
      if (consolidatedCoordinates.length === 0 || j > 0) {
        consolidatedCoordinates.push(segCoords[j]);
      }
    }

    const fromNode = nodeMap.get(dEdge.fromNodeId)!;
    const toNode = nodeMap.get(dEdge.toNodeId)!;
    const effectiveSpeed = Math.max(5, edge.base_speed_kmh / Math.max(1, dEdge.multiplier));

    segments.push({
      edge_id: edge.id,
      name: edge.name,
      from_node_id: dEdge.fromNodeId,
      to_node_id: dEdge.toNodeId,
      from_node_name: fromNode.name,
      to_node_name: toNode.name,
      distance_m: edge.distance_m,
      base_speed_kmh: edge.base_speed_kmh,
      effective_speed_kmh: Number(effectiveSpeed.toFixed(1)),
      travel_time_s: Math.round(dEdge.effectiveTravelTimeS),
      flood_depth_cm: Number(dEdge.depth_cm.toFixed(1)),
      risk_level: dEdge.risk_level,
      penalty_factor: Number(dEdge.multiplier.toFixed(2)),
      is_blocked: dEdge.isBlocked,
      coordinates: segCoords,
    });
  }

  // Filter avoided hazards to only those not on the selected path
  const relevantAvoided = allAvoidedHazards.filter((h) => !traversedEdgeIds.has(h.edge_id));

  return {
    status: 'FOUND',
    mode,
    horizon,
    origin,
    destination,
    total_distance_m,
    total_distance_km: Number((total_distance_m / 1000).toFixed(2)),
    total_time_s: Math.round(total_time_s),
    total_time_min: Number((total_time_s / 60).toFixed(1)),
    max_flood_depth_cm: Number(max_flood_depth_cm.toFixed(1)),
    highest_risk_level,
    segments,
    path_node_ids: nodeSequence,
    coordinates: consolidatedCoordinates,
    risk_exposure,
    avoided_segments: relevantAvoided,
    explanation: generateRouteExplanation(
      mode,
      'FOUND',
      max_flood_depth_cm,
      highest_risk_level,
      relevantAvoided,
      segments
    ),
    is_alternative: isAlternative,
    provenance: 'MODEL OUTPUT / DERIVED',
  };
}

/**
 * Computes primary flood-safe route and optional alternative route between origin and destination.
 */
export function findRoute(
  originId: string,
  destinationId: string,
  mode: RoutingMode,
  horizon: 'T+0' | 'T+1' | 'T+2' | 'T+3',
  riskForecast: RiskForecast,
  graph: RoadNetworkGraph = PROTOTYPE_ROAD_GRAPH
): RoutePairResult {
  const nodeMap = new Map<string, RoadNode>(graph.nodes.map((n) => [n.id, n]));
  const origin = nodeMap.get(originId);
  const destination = nodeMap.get(destinationId);

  if (!origin || !destination) {
    throw new Error(`Invalid origin (${originId}) or destination (${destinationId})`);
  }

  const horizonState = riskForecast.horizons[horizon];
  const assessments = horizonState ? horizonState.cellAssessments : [];

  // Primary route calculation
  const primarySolve = dijkstra(originId, destinationId, graph, assessments, mode);
  const primary = formatRouteResult(
    origin,
    destination,
    mode,
    horizon,
    primarySolve.path,
    primarySolve.nodeSequence,
    primarySolve.allAvoidedHazards,
    nodeMap,
    false
  );

  let alternative: RouteResult | undefined = undefined;

  // If primary route exists with at least 1 edge, attempt to compute a distinct alternative route
  if (primary.status === 'FOUND' && primary.segments.length > 0) {
    // Penalize the edges used by the primary route to encourage finding a viable detour
    const primaryEdgeIds = new Set(primary.segments.map((s) => s.edge_id));
    const altSolve = dijkstra(originId, destinationId, graph, assessments, mode, primaryEdgeIds);

    if (altSolve.success && altSolve.path.length > 0) {
      const altResult = formatRouteResult(
        origin,
        destination,
        mode,
        horizon,
        altSolve.path,
        altSolve.nodeSequence,
        altSolve.allAvoidedHazards,
        nodeMap,
        true
      );

      // Verify alternative is physically distinct (not identical edge sequence)
      const altEdgeIds = altResult.segments.map((s) => s.edge_id).join('-');
      const priEdgeIds = primary.segments.map((s) => s.edge_id).join('-');
      if (altEdgeIds !== priEdgeIds) {
        alternative = altResult;
      }
    }
  }

  return {
    primary,
    alternative,
    evaluated_at: new Date().toISOString(),
    horizon,
    mode,
    provenance: 'MODEL OUTPUT / DERIVED',
  };
}

File: backend/app/services/routing_service.py
import math
import networkx as nx
from datetime import datetime
from typing import Dict, Any, List, Tuple
from app.data.seed_data import MUMBAI_ROAD_NODES, MUMBAI_ROAD_EDGES
from app.services.risk_service import get_risk_classification

def haversine_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    R = 6371000.0
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dp, dl = math.radians(lat2 - lat1), math.radians(lng2 - lng1)
    a = math.sin(dp / 2.0) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dl / 2.0) ** 2
    return R * (2.0 * math.atan2(math.sqrt(a), math.sqrt(1.0 - a)))

def find_nearest_node(lat: float, lng: float) -> Dict[str, Any]:
    best = MUMBAI_ROAD_NODES[0]
    min_d = float('inf')
    for n in MUMBAI_ROAD_NODES:
        d = haversine_distance(lat, lng, n["lat"], n["lng"])
        if d < min_d:
            min_d = d
            best = n
    return best

def build_road_network_graph() -> nx.Graph:
    G = nx.Graph()
    for n in MUMBAI_ROAD_NODES:
        G.add_node(n["id"], name=n["name"], lat=n["lat"], lng=n["lng"], type=n.get("type", "JUNCTION"), catchment_cell_id=n.get("catchment_cell_id", "DEM-GRID-1-1"))
    for e in MUMBAI_ROAD_EDGES:
        dist = e["distance_m"]
        speed = e["base_speed_kmh"]
        travel_time = dist / (speed / 3.6) if speed > 0 else 60.0
        G.add_edge(
            e["source"], e["target"],
            id=e["id"], name=e["name"], distance_m=dist, base_speed_kmh=speed,
            base_travel_time_s=travel_time, road_type=e.get("road_type", "ARTERIAL"),
            is_elevated=e.get("is_elevated", False), coordinates=e["coordinates"]
        )
    return G

def get_edge_depth(edge_data: Dict[str, Any], rainfall_mm_hr: float = 48.5) -> float:
    if edge_data.get("is_elevated", False) or rainfall_mm_hr <= 0:
        return 0.0
    scale = rainfall_mm_hr / 48.5
    edge_id = edge_data.get("id", "")
    edge_name = edge_data.get("name", "")
    if "Hindmata" in edge_name or edge_id == "RE-02":
        return round(120.0 * scale, 1)
    elif "Kurla" in edge_name or edge_id == "RE-07":
        return round(105.0 * scale, 1)
    elif "BKC-Kurla" in edge_name or edge_id == "RE-09":
        return round(45.0 * scale, 1)
    elif "Dadar-Wadala" in edge_name or edge_id == "RE-05":
        return round(18.0 * scale, 1)
    else:
        return 0.0

def compute_safe_route_pair(
    origin_lat: float,
    origin_lng: float,
    dest_lat: float,
    dest_lng: float,
    routing_mode: str = "SAFEST",
    vehicle_type: str = "EMERGENCY_AMBULANCE",
    max_allowable_depth_cm: float = 20.0,
    rainfall_mm_hr: float = 48.5
) -> Dict[str, Any]:
    G = build_road_network_graph()
    orig = find_nearest_node(origin_lat, origin_lng)
    dest = find_nearest_node(dest_lat, dest_lng)
    orig_id, dest_id = orig["id"], dest["id"]

    # 1. Standard Route (Shortest time)
    std_path = nx.shortest_path(G, source=orig_id, target=dest_id, weight="base_travel_time_s")
    
    # 2. Penalty Safe Route
    def safe_weight(u, v, e):
        base_t = e.get("base_travel_time_s", 60.0)
        depth = get_edge_depth(e, rainfall_mm_hr)
        if depth > max_allowable_depth_cm:
            penalty = 1.0 + 100.0 * ((depth - max_allowable_depth_cm) ** 2)
            return base_t * penalty
        return base_t

    safe_path = nx.dijkstra_path(G, source=orig_id, target=dest_id, weight=safe_weight)

    def build_result(path_nodes: List[str], mode_label: str) -> Dict[str, Any]:
        segs = []
        tot_dist, tot_time = 0.0, 0.0
        max_d = 0.0
        highest_risk = "LOW"
        coords = []
        avoided = []
        
        for i in range(len(path_nodes) - 1):
            u, v = path_nodes[i], path_nodes[i+1]
            e = G[u][v]
            dist = e.get("distance_m", 1000.0)
            speed = e.get("base_speed_kmh", 40.0)
            depth = get_edge_depth(e, rainfall_mm_hr)
            risk_info = get_risk_classification(depth)
            
            if depth > max_allowable_depth_cm:
                eff_speed = 5.0
                blocked = True
                avoided.append({
                    "edge_id": e["id"],
                    "name": e["name"],
                    "risk_level": risk_info["risk_level"],
                    "depth_cm": depth,
                    "reason": f"Depth {depth} cm exceeds allowable threshold ({max_allowable_depth_cm} cm)"
                })
            else:
                eff_speed = speed
                blocked = False
                
            t_s = dist / (eff_speed / 3.6)
            tot_dist += dist
            tot_time += t_s
            if depth > max_d:
                max_d = depth
                highest_risk = risk_info["risk_level"]
                
            edge_coords = e.get("coordinates", [[G.nodes[u]["lat"], G.nodes[u]["lng"]], [G.nodes[v]["lat"], G.nodes[v]["lng"]]])
            if i == 0:
                coords.extend(edge_coords)
            else:
                coords.extend(edge_coords[1:])
                
            segs.append({
                "edge_id": e["id"],
                "name": e["name"],
                "from_node_id": u,
                "to_node_id": v,
                "from_node_name": G.nodes[u]["name"],
                "to_node_name": G.nodes[v]["name"],
                "distance_m": round(dist, 1),
                "base_speed_kmh": speed,
                "effective_speed_kmh": round(eff_speed, 1),
                "travel_time_s": round(t_s, 1),
                "flood_depth_cm": depth,
                "risk_level": risk_info["risk_level"],
                "penalty_factor": 1.0 if not blocked else 15.0,
                "is_blocked": blocked,
                "coordinates": edge_coords
            })

        return {
            "status": "FOUND",
            "mode": mode_label,
            "horizon": "T+1",
            "origin": {
                "id": orig["id"], "name": orig["name"], "lat": orig["lat"], "lng": orig["lng"],
                "type": orig.get("type", "JUNCTION"), "catchment_cell_id": orig.get("catchment_cell_id", "DEM-GRID-1-1")
            },
            "destination": {
                "id": dest["id"], "name": dest["name"], "lat": dest["lat"], "lng": dest["lng"],
                "type": dest.get("type", "JUNCTION"), "catchment_cell_id": dest.get("catchment_cell_id", "DEM-GRID-1-1")
            },
            "total_distance_m": round(tot_dist, 1),
            "total_distance_km": round(tot_dist / 1000.0, 2),
            "total_time_s": round(tot_time, 1),
            "total_time_min": round(tot_time / 60.0, 1),
            "max_flood_depth_cm": max_d,
            "highest_risk_level": highest_risk,
            "segments": segs,
            "path_node_ids": path_nodes,
            "coordinates": coords,
            "risk_exposure": {"low_m": tot_dist if max_d < 5 else 0.0, "moderate_m": 0, "high_m": 0, "very_high_m": 0, "critical_m": tot_dist if max_d >= 100 else 0},
            "avoided_segments": avoided,
            "explanation": f"Computed {mode_label} path. Max flood depth along path: {max_d} cm.",
            "is_alternative": mode_label == "FASTEST"
        }

    primary = build_result(safe_path, routing_mode.upper())
    alternative = build_result(std_path, "FASTEST")

    return {
        "status": "success",
        "primary": primary,
        "alternative": alternative,
        "evaluated_at": datetime.utcnow().isoformat() + "Z",
        "horizon": "T+1",
        "mode": routing_mode.upper()
    }
